computation-local/src/main/java/com/powsybl/computation/local/LocalComputationManager.java

```diff
@@ -340,22 +347,37 @@ private void exit() {
         return f;
     }
 
-    private class MyCf<R> extends ThreadInterruptedCompletableFuture<R> {
+    private class LocalCompletableFuture<R> extends ThreadInterruptedCompletableFuture<R> {
 
         private Path workingDir;
+        private volatile boolean cancel = false;
 
         private void setWorkingDir(Path workingDir) {
             this.workingDir = workingDir;
         }
 
+        /**
+         * Throws a {@link CancellationException} if cancel has been requested.
+         */
+        private synchronized void checkCancellation() {
+            if (isCancelled()) {
+                throw new CancellationException("Cancellation before starting the command.");
+            }
+        }
+
         @Override
         public boolean cancel(boolean mayInterruptIfRunning) {
-            super.cancel(mayInterruptIfRunning);
+            cancel = true;
```

Comment:
Need opinions about this change, @geofjamg , @yichen88 , @mathbagu , @jonenst .

Now we complete the future with a `CancellationException` as specified in the class API, but do you agree to remove the thread interrupt ?

I think the interrupt may be more misleading than really useful, unless there are some really heavy processing in `before` or `after` methods, therefore I think we can remove it. In this case we should not use anymore the `ThreadInterruptedCompletableFuture` as a super class. Also, this implementation of cancel does not propagate the exception to dependent, whereas the standard implementation does : maybe we should call instead. So, we could simply have :
```java
private class LocalCompletableFuture<R> extends CompletableFuture<R> {

        @Override
        public boolean cancel(boolean mayInterruptIfRunning) {
            super.cancel(mayInterruptIfRunning);
            if (mayInterruptIfRunning) {
                localCommandExecutor.stop(workingDir);
            }
            return true;
        }
    }
```
(Note: in this case we don't need anymore to explicitly throw CancellationException in the code above).

Another possibility is to leave the interrupt but only after we complete with `CancellationException`, this way the user of the future will only see this exception, but the other processing will still be interrupted behind the scene.

Reply:
Maybe an even simpler and better approach is to not subclass `CompletableFuture` at all, and simply chain the "computation" future  with:
```java
return f.exceptionally(/* method which kills the underlying process. */)
```

Reply:
Again another possibility is to not subclass `CompletableFuture`, but to chain the returned future with an exception handling method:
```java
f.whenComplete((r, ex) -> { /* Kill underlying process if cancellation exception */ })
return f;
```

Reply:
We would need feedback/opinions about this change, @geofjamg , @yichen88 , @mathbagu , @jonenst .

Reply:
I'm not sure my answer is relevant, but I think we have to call the after() method, if the cancellation is performed during the execution, and skip it if the task is cancelled before its startup. The `after` method need to know the status of the execution (cancelled or finished)

There is probably a thread safety issue with the attribute cancel, the synchronized method `checkCancellation`, and the `isCancelled` method. Maybe we should use an `AtomicBoolean`.

Reply:
For the concurrent issue, i think we could even remove `synchronized` in `checkCancellation`. As there is just one thread which call cancel() and modify the boolean value and others could READ the UPDATED value.

Reply:
Thank you for the feedback. I just proposed an implementation and added unit tests to demonstrate the behaviour on cancellation.

In this proposition, we use the plain `CompletableFuture` class.
 * A callback is simply added to stop the command execution when the future completes with `CancellationException`.
 * There is no more thread interruption.
 * The `after` method is not called anymore when cancellation is requested, which could be changed in a further development where the post-processing should have the information that the computation has been cancelled.
 * There isn't more synchronization on cancellation than what is proposed by the `CompletableFuture` implementation, which may be non-optimal : in some corner cases, command execution will not be properly killed, see comment in the code

Reply:
Following last feedback, namely that the `cancel` should still interrupt the thread, and that we may use the behaviour of `Future` class to implement this, I just proposed a new attempt at solving this issue.

This new implementation proposes:  
* a reusable implementation of `CompletableFuture` bound to an actual task,
which overrides the behaviour of `cancel` to interrupt the task thread.
It re-uses the JDK FutureTask class, which does just this and has better behaviour
than the previous implementation (no waiting when thread is not set ...)
* the use of such a CompletableFuture in the local computation manager. It now relies on the interrupt to decide when the underlying process should also be killed. Note: actually this could even be handled at lower level by the `LocalCommandExecutor`.

Does it seem a better solution ?

Reply:
I prefer the implementation based on not subclassing CompletableFuture, but chaining the returned future with an exception handling method. Let met write some code to demonstrate what I have in mind

Reply:
Ok, any improvement will be welcome !

I would have preferred too, but finally it did not seem that obvious to make it concise and reusable. The proposed solution encapsulates well the solution in a reusable class (I have powsybl-slurm in mind) : I think the alternative solution should propose such a class or a utility method.

In any case we at least have one working solution now, so it would be nice not to postpone too much the resolution.